src/services/fireblocks.service.js

```diff
@@ -0,0 +1,369 @@
+      continue;
+    }
+    const addressInfo = await storage.getAddressInfo(spentOut.decoded.address);
+    if (!addressInfo) {
+      // Not a wallet address
+      continue;
+    }
+
+    const pubkeyHex = await storage.getAddressPubkey(addressInfo.bip32AddressIndex);
+    const pubkey = Buffer.from(pubkeyHex, 'hex');
+
+    signingIndices.push({
+      inputIndex,
+      addressIndex: addressInfo.bip32AddressIndex,
+      pubkey,
+    });
+  }
+
+  /* istanbul ignore next */
+  if (tx.version === hathorLib.constants.NANO_CONTRACTS_VERSION) {
+    const { pubkey } = tx;
+    const address = hathorLib.addressUtils.getAddressFromPubkey(pubkey.toString('hex'), storage.config.getNetwork());
+    const addressInfo = await storage.getAddressInfo(address.base58);
+    if (addressInfo) {
+      ncCallerIndex = addressInfo.bip32AddressIndex;
+    }
+  }
+
+  // Now we sign
+  const indices = signingIndices.map(data => data.addressIndex);
+  if (ncCallerIndex) {
+    indices.push(ncCallerIndex);
+  }
+
+  // create signature data object from the response
+  const fbTxInfo = await client.sendRawTransaction(dataToSignHash, indices);
+  if (!fbTxInfo.signedMessages) {
+    throw new Error(`Fireblocks did not return signatures for the transaction(${fbTxInfo.id}); external-id: (${dataToSignHash.toString('hex')})`);
+  }
+  const signedMessagesObj = {};
+  for (const item of fbTxInfo.signedMessages) {
+    if (item.content !== dataToSignHash.toString('hex')) {
+      throw new Error(`Fireblocks returned inconsistent signatures in transaction(${fbTxInfo.id}); external-id: (${dataToSignHash.toString('hex')})`);
+    }
+    if (item.derivationPath[4] in signedMessagesObj) {
+      // duplicate signature due to inputs from the same address
+      // Since the content is the same we can safely skip it
+      continue;
+    }
+    signedMessagesObj[item.derivationPath[4]] = {
+      signature: fireblocksSignatureToDER(item.signature),
+      content: item.content,
+      publicKey: item.publicKey,
+    };
+  }
+
+  for (const data of signingIndices) {
+    const sig = signedMessagesObj[data.addressIndex];
+    if (!sig) {
+      throw new Error(`Fireblocks did not return signature for address index ${data.addressIndex} in transaction(${fbTxInfo.id}); external-id: (${dataToSignHash.toString('hex')})`);
+    }
+    if (sig.publicKey !== data.pubkey.toString('hex')) {
+      throw new Error(`Fireblocks signature does not match locally generated public key in transaction(${fbTxInfo.id}); external-id: (${dataToSignHash.toString('hex')})`);
+    }
+    inputSignatures.push({
+      ...data,
+      signature: sig.signature,
+    });
+  }
+
+  if (ncCallerIndex) {
+    const sig = signedMessagesObj[ncCallerIndex];
+    if (!sig) {
+      throw new Error(`Fireblocks did not return signature for address index ${ncCallerIndex} in transaction(${fbTxInfo.id}); external-id: (${dataToSignHash.toString('hex')})`);
+    }
+    ncCallerSignature = sig.signature.fullSig;
```

Comment:
Why is it different for the nc signature?

In the others you use `fireblocksSignatureToDER(item.signature)` and here `sig.signature.fullSig`

Reply:
It should not be different, fixed.